### PR TITLE
Content types

### PR DESCRIPTION
**Pull request by tvh**
My take on the change to support https://github.com/haskell-servant/servant-server/pull/13. The generic changes necessary beforehand are in https://github.com/haskell-servant/servant-client/pull/10.

This is the simplest approach just picking the first element.

### CODE REVIEW

**Comment by alpmestan** [comment]
Hey,

This all looks great. I just have one comment though: wouldn't we want to be able to pick the content type  at the use-site? Providing a `Proxy :: Proxy JSON` argument for JSON, things like that?

**Comment by tvh** [comment]
I would be nice, yes. Unfortunately I couldn't find a way to both support that AND provide an interface for people not interested in that. Either one of these was fairly simple though.

**Comment by christian-marie** [comment]
I couldn't find a clean approach either

File: src/Servant/Client.hs [hunk 1]
```diff
@@ -436,14 +439,14 @@ instance HasClient Raw where
...
+      => HasClient (ReqBody (ct ': cts) a :> sublayout) where
 
-  type Client (ReqBody a :> sublayout) =
+  type Client (ReqBody (ct ': cts) a :> sublayout) =
     a -> Client sublayout
 
   clientWithRoute Proxy req body =
     clientWithRoute (Proxy :: Proxy sublayout) $
-      setRQBody (encode body) req
+      setRQBody (encode body) ("application" // "json" /: ("charset", "utf-8")) req
```
Comment (jkarni) [inline src/Servant/Client.hs]: `charset` is for `text`, not `application`. (This was my mistake - I did this in `servant`.)
I think this should fix the failing test cases, and then we're good to go.

**Comment by tvh** [comment]
I updated this to also test with multiple content-types. The reason why the tests fail is: decoding of JSON used to be done with a `decodeLenient`, which also allows non-root JSON things to be parsed.

**Comment by jkarni** [comment]
@tvh I see. I'll use `decodeLenient` as the `fromByteString` implementation in `servant`.

**Comment by jkarni** [comment]
All right! Thanks a lot!

**Pull request merged**